syntax = "proto3";
package sports_keypoints_proto;

import "common.proto";

service GolfKeypointsService {
    rpc uploadInputImage(UploadInputImageRequest) returns (UploadInputImageResponse) {}
    rpc listInputImagesForUser(ListInputImagesForUserRequest) returns (ListInputImagesForUserResponse) {}
    rpc readInputImage(ReadInputImageRequest) returns (ReadInputImageResponse) {}
    rpc deleteInputImage(DeleteInputImageRequest) returns (DeleteInputImageResponse) {}

    // optional if want specific keypoints
    rpc calibrateInputImage(CalibrateInputImageRequest) returns (CalibrateInputImageResponse) {}

    rpc calculateGolfKeypoints(CalculateGolfKeypointsRequest) returns (CalculateGolfKeypointsResponse) {}
    rpc readGolfKeypoints(ReadGolfKeypointsRequest) returns (ReadGolfKeypointsResponse) {}
    // if estimated body keypoints are off or have low confidence, client can manually input where body parts are
    rpc updateBodyKeypoints(UpdateBodyKeypointsRequest) returns (UpdateBodyKeypointsResponse) {}
    rpc deleteGolfKeypoints(DeleteGolfKeypointsRequest) returns (DeleteGolfKeypointsResponse) {}

    // TODO: Stream for videos
}

// TODO: Add name and desc tags for input image (ie. june-24th, 2020, lift left foot in backswing)
message UploadInputImageRequest {
    string session_token = 1;
    ImageType image_type = 2;
    bytes image = 3;
}

message UploadInputImageResponse {
    bool success = 1;
    string input_image_id = 2;
}

message ListInputImagesForUserRequest{
    string session_token = 1;
}

message ListInputImagesForUserResponse {
    bool success = 1;
    repeated string input_image_ids = 2;
}

message ReadInputImageRequest {
    string session_token = 1;
    string input_image_id = 2;
}

message ReadInputImageResponse {
    bool success = 1;
    ImageType image_type = 2;
    bytes image = 3;
    CalibrationType calibration_type = 4;
    FeetLineMethod feet_line_method = 5;
}

message DeleteInputImageRequest {
    string session_token = 1;
    string input_image_id = 2;
}

message DeleteInputImageResponse {
    bool success = 1;
}

message CalibrateInputImageRequest {
    string session_token = 1;
    string input_image_id = 2;
    CalibrationType calibration_type = 3;
    FeetLineMethod feet_line_method = 4;
    // only required if want certain data for DTL and Face On
    bytes calibration_image_axes = 5;
    // only required if want certain data for DTL
    bytes calibration_image_vanishing_point = 6;
    // only required if want certain data such as distance from ball and ball position
    Keypoint golf_ball = 7;
    // only required if want certain data such as ulnar deviation and shaft lean (club_head is also required)
    Keypoint club_butt = 8;
    // only required if want certain data such as ulnar deviation and shaft lean (club_butt is also required)
    Keypoint club_head = 9;
    // optional: horizontal line and vertical line points for axes calibration
    // optional: parallel line points for vanishing point
}

message CalibrateInputImageResponse {
    bool success = 1;
}

message CalculateGolfKeypointsRequest {
    string session_token = 1;
    string input_image_id = 2;
}

message CalculateGolfKeypointsResponse {
    bool success = 1;
    bytes output_image = 2;
    GolfKeypoints golf_keypoints = 3;
}

message ReadGolfKeypointsRequest {
    string session_token = 1;
    string input_image_id = 2;
}

message ReadGolfKeypointsResponse {
    bool success = 1;
    bytes output_image = 2;
    GolfKeypoints golf_keypoints = 3;
}

message UpdateBodyKeypointsRequest {
    string session_token = 1;
    string input_image_id = 2;
    Body25PoseKeypoints updated_body_keypoints = 3;
}

message UpdateBodyKeypointsResponse {
    bool success = 1;
    GolfKeypoints updated_golf_keypoints = 2;
}

message DeleteGolfKeypointsRequest {
    string session_token = 1;
    string input_image_id = 2;
}

message DeleteGolfKeypointsResponse {
    bool success = 1;
}

enum ImageType {
    IMAGE_TYPE_UNSPECIFIED = 0; 
    FACE_ON = 1;
    DTL = 2;
}

enum CalibrationType {
    NO_CALIBRATION = 0;
    AXES_CALIBRATION_ONLY = 1;
    // vanishing point requires axes calibration as well, same as full for Dtl, will just do axes if faceon
    AXES_AND_VANISHING_POINT_CALIBRATION = 2; 
    // axes for Faceon, axes and vanishing point for Dtl
    FULL_CALIBRATION = 3; 
}

enum FeetLineMethod {
    // will default to use heel line
    FEET_LINE_METHOD_UNSPECIFIED = 0; 
    USE_HEEL_LINE = 1;
    USE_TOE_LINE = 2;
}

message GolfKeypoints {
    DTLGolfSetupPoints dtl_golf_setup_points = 1;
    FaceOnGolfSetupPoints faceon_golf_setup_points = 2;
    Body25PoseKeypoints body_keypoints = 3;
}

message DTLGolfSetupPoints {
    // degrees from vertical, requires axes calibration
    Double spine_angle = 1;
    // degrees from target (negative is open, positive is closed), requires axes and vanishing point calibration
    Double feet_alignment = 2;
    // degrees from target (negative is open, positive is closed), requires axes and vanishing point calibration
    Double heel_alignment = 3;
    // degrees from target (negative is open, positive is closed), requires axes and vanishing point calibration
    Double toe_alignment = 4;
    // degrees between line from hip to knee and line from knee to ankle, for best results wear shorts/tighter pants
    Double knee_bend = 5;
    // degrees from target (negative is open, positive is closed), requires axes and vanishing point calibration (note: shoulder alignment is very sensitive to detection, threshold should be higher for shoulder keypoint detection for accurate results)
    Double shoulder_alignment = 6;
    //waist_alignment, requires axes and vanishing point calibration
    //distance_from_ball // requires golf ball 
    //ulnar_deviation // requires club butt and club head
    //waist_bend: (same as spine angle??)
    //neck_angle
    //spine_bend
}

message FaceOnGolfSetupPoints {
    // degrees from vertical (positive is right side bend, negative is left side bend), requires axes calibration
    Double side_bend = 1;
    // degrees from line running through midpoint of heels, perpendicular to target line (positive external feet, negative is internal feet), requires axes calibration
    Double l_foot_flare = 2;
    // degrees from line running through midpoint of heels, perpendicular to target line (positive external feet, negative is internal feet), requires axes calibration
    Double r_foot_flare = 3;
    // ratio of stance width to shoulder width, greater than 1 is wider than shoulders, less than 1 is less wide than shoulders
    Double stance_width = 4;
    // degrees offset from horizontal axis, positive is trail shoulder lower, negative is lead shoulder lower, requires axes calibration
    Double shoulder_tilt = 5; 
    //waist_angle, requires axes calibration
    //shaft_lean // requires club butt and club head
    //ball_position (relative to feet and chest and head?) // requires golf ball
    //head_position
    //chest_position
}
